### PR TITLE
[Catlin] Add a script linter command

### PR DESCRIPTION
**Pull request by chmouel**
# Changes

This add a command to catlin to launch a language linter on a script field :

`catlin lint script`

Assuming a task as argument it will go over all the tasks and validate according
to a linter.

Configuration is pretty static and in code at the moment, and only has linter
for posix shell/bash and python, will expand on this in further version to be
able to provide a configuration file.

This doesn't handle TaskSpec from taskRuns and pipeline/pipelineRuns

Output would look like this : 

```sh
$ catlin lint script $GOPATH/src/github.com/tektoncd/catalog/task/*/*/*.yaml
WARN : step: ansible-runner is not using #!/usr/bin/env 
WARN : Couldn't find the linter shellcheck in the path
WARN : Couldn't find the linter shellcheck in the path
WARN : Couldn't find the linter shellcheck in the path
WARN : Couldn't find the linter shellcheck in the path
WARN : Couldn't find the linter shellcheck in the path
WARN : Couldn't find the linter shellcheck in the path
WARN : Couldn't find the linter shellcheck in the path
WARN : Couldn't find the linter shellcheck in the path
WARN : Couldn't find the linter shellcheck in the path
WARN : Couldn't find the linter shellcheck in the path
WARN : Couldn't find the linter shellcheck in the path
WARN : Couldn't find the linter shellcheck in the path
WARN : step: git-clone is not using #!/usr/bin/env 
WARN : Couldn't find the linter shellcheck in the path
WARN : step: github-add-comment is not using #!/usr/bin/env 
ERROR: /usr/bin/pylint, [-dC0103] failed:
************* Module catlin-script-linter501955988
github-add-comment-add-comment:27:0: W0311: Bad indentation. Found 2 spaces, expected 4 (bad-indentation)
github-add-comment-add-comment:29:0: W0311: Bad indentation. Found 2 spaces, expected 4 (bad-indentation)
github-add-comment-add-comment:44:0: W0311: Bad indentation. Found 2 spaces, expected 4 (bad-indentation)
github-add-comment-add-comment:1:0: C0114: Missing module docstring (missing-module-docstring)

-----------------------------------
Your code has been rated at 7.89/10


WARN : step: github-add-comment is not using #!/usr/bin/env 
ERROR: /usr/bin/pylint, [-dC0103] failed:
************* Module catlin-script-linter971370467
github-add-comment-add-comment:27:0: W0311: Bad indentation. Found 2 spaces, expected 4 (bad-indentation)
github-add-comment-add-comment:29:0: W0311: Bad indentation. Found 2 spaces, expected 4 (bad-indentation)
github-add-comment-add-comment:44:0: W0311: Bad indentation. Found 2 spaces, expected 4 (bad-indentation)
github-add-comment-add-comment:1:0: C0114: Missing module docstring (missing-module-docstring)

-----------------------------------
Your code has been rated at 7.89/10
[...]
```

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

File: catlin/pkg/linter/script.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+type config struct {
+	regexp  string
+	linters []linter
+}
+
+// NewConfig construct default config
+func NewConfig() []config {
+	return []config{
```
Comment (vdemeester) [inline catlin/pkg/linter/script.go:46]: For a script with no shebang `#!` we should also run shellcheck and `/bin/sh -n` 🙃

**Review by vdemeester — APPROVED** [review]

File: catlin/pkg/linter/script.go (hunk 1 above)
Comment (chmouel) [reply catlin/pkg/linter/script.go:46]: yeah good point, i have added it in the tests but forgot to bring it back to the main config,

Comment (chmouel) [reply catlin/pkg/linter/script.go:46]: nm my previous comment i do have it: 

https://github.com/tektoncd/plumbing/pull/679/files/0c8f582d16de69ab65fe52c44d7b100d0339c8ef#diff-bf727d57cc66d19dbee8015f9adcddab76e0d36677f82b2dd348ca237ec7e181R89-R92

File: catlin/pkg/linter/script.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+			}},
+		config{regexp: `(/usr/bin/env\s|.*/bin/|/usr/libexec/platform-)python(23)?`,
+			linters: []linter{
+				linter{
+					cmd:  "pylint",
+					args: []string{"-dC0103"}, // Disabling C0103 which is invalid name convention
+				},
+			}},
+	}
+	// }
```
Comment (vdemeester) [inline catlin/pkg/linter/script.go:78]: Shouldn't this be removed 😝

File: catlin/pkg/linter/script.go (hunk 1 above)
Comment (vdemeester) [reply catlin/pkg/linter/script.go:46]: ah right 😅

**Review by sthaha — COMMENTED** [review]
/lgtm
/approve

This is really nice, @chmouel !

File: catlin/.gitignore [hunk 3]
```diff
@@ -0,0 +1 @@
+catlin
```
Comment (savitaashture) [inline catlin/.gitignore:1]: any reason to ignore `catlin` here :thinking:

File: catlin/pkg/linter/script.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
+// Copyright © 2020 The Tekton Authors.
```
Comment (savitaashture) [inline catlin/pkg/linter/script.go:1]:
```suggestion
// Copyright © 2021 The Tekton Authors.
```

**Review by savitaashture — COMMENTED** [review]
looks good to me :+1: 

few minor comments

File: catlin/.gitignore (hunk 3 above)
Comment (vdemeester) [reply catlin/.gitignore:1]: to make sure we don't commit a binary we compiled while developing on it I think 😉

Comment (chmouel) [reply catlin/.gitignore:1]: that's correct :-p 

I actually have another followup pr that adds a proper makefile with linting and generation etc..

Comment (savitaashture) [reply catlin/.gitignore]: :+1:

**Comment by savitaashture** [comment]
/lgtm

Thank you @chmouel :+1:

**Review by vdemeester — APPROVED** [review]
/meow

**Pull request merged**